client/jetpack-cloud/controller.ts

```diff
@@ -0,0 +1,217 @@
+import type PageJS from 'page';
+
+/**
+ * Parse site slug from path. If no slug is detected but a `site` query parameter
+ * exists, a redirection to `/:path/:site/` occurs.
+ *
+ * @param {PageJS.Context} context Route context
+ * @returns {string} Site slug
+ */
+const parseSiteFragment = ( context: PageJS.Context ): string | undefined => {
+	const siteFragment = context.params.site || getSiteFragment( context.path );
+
+	if ( siteFragment ) {
+		return siteFragment;
+	}
+
+	const { query: queryParams, pathname } = context;
+	const { site: siteQuery } = queryParams;
+
+	if ( siteQuery ) {
+		page.redirect(
+			addQueryArgs(
+				{
+					...queryParams,
+					site: undefined,
+				},
+				`${ pathname }/${ siteQuery }`
+			)
+		);
+	}
+};
+
+/**
+ * Fetch site data.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {number|string} siteId Site id
+ * @returns {Promise} Promise that resolves with the site id and slug
+ */
+const fetchSite = (
+	context: PageJS.Context,
+	siteId: number | string
+): Promise< { id: number | undefined; slug: string | undefined } > => {
+	const { getState, dispatch } = context.store;
+
+	return dispatch( requestSite( siteId ) ).then( () => ( {
+		id: getSiteId( getState(), siteId ),
+		slug: getSiteSlug( getState(), siteId ),
+	} ) );
+};
+
+/**
+ * Store site id in state, and add site to the list of most recent sites.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {number|string} siteId Site id
+ */
+const selectSite = ( context: PageJS.Context, siteId: number ): void => {
+	const { dispatch } = context.store;
+
+	dispatch( setSelectedSiteId( siteId ) );
+	updateRecentSitesPreferences( context );
+};
+
+/**
+ * Handle case when path contains no site slug. If the current user has only
+ * one site, we try to use this site for context. Otherwise, the user must
+ * select one of their sites.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {Function} next Next middleware function
+ */
+const siteSelectionWithoutFragment = ( context: PageJS.Context, next: () => void ): void => {
+	const { getState, dispatch } = context.store;
+	const currentUser = getCurrentUser( getState() ) as UserData;
+	const hasOneSite = currentUser?.visible_site_count === 1;
+
+	// User has only one site
+	if ( hasOneSite ) {
+		const primarySiteId = getPrimarySiteId( getState() ) as NonNullable< number >;
+		const primarySiteSlug = getSiteSlug( getState(), primarySiteId );
+
+		// Site is already available in state
+		if ( primarySiteSlug ) {
+			// Redirect to the single site context
+			redirectToPrimary( context, primarySiteSlug );
+		} else {
+			fetchSite( context, primarySiteId ).then( ( { slug } ) => {
+				if ( slug ) {
+					// Redirect to the single site context
+					redirectToPrimary( context, slug );
+				} else {
+					// Display error notification
+					showMissingPrimaryError( currentUser, dispatch );
+				}
+			} );
+		}
+	} else {
```

Comment:
Have you tested this with the edge case when the user has no sites at all?

Reply:
We're planning on handling that in a separate PR that checks whether the user has WP.com sites and redirects or not accordingly, so that section of this is TBD.